KEP-3203: Add Auto-refreshing Official CVE feed (#3204)
kubernetes/enhancements: keps/sig-security/3203-auto-refreshing-official-cve-feed/README.md

As a K8s End User, I want to use my browser to get a list of fixed CVEs, from
the official K8s website so that I can trust it as an authoritative source of
data through implicit trust offered by website certificate and domain name.

#### Story 3

As a K8s maintainer, I want to create a process that auto-updates CVE feed, when
SRC announces new CVEs such that I do not have to do extra work to maintain this
feed manually

## Proposal

### Pre-requisites

- [x] Add official-cve-label https://github.com/kubernetes/test-infra/pull/23428
- [x] Search and Identify closed issues that have a CVE ID e.g. CVE-1001-12345
  in the issue description or summary (This
  search [filter](https://github.com/kubernetes/kubernetes/issues?q=is%3Aissue+in%3Abody+%22CVSS%3A3.%22+label%3Acommittee%2Fsecurity-response+is%3Aclosed+)
  is giving the most accurate data so far)
- [x] Label those issues with `official-cve-feed`
  using https://docs.github.com/en/rest/reference/issues REST API
- [x] Add `official-cve-feed` label as part of SRC
  playbook: https://github.com/kubernetes/committee-security-response/pull/133

### Overview

- Generate a JSON blob using the results from the filtered label on `k/k`

>   repo.

Reviewer: What about for non-k/k projects? E.g. https://github.com/kubernetes/ingress-nginx/issues/7837

- Create a Prow job to periodically generate this JSON blob.
- Push this JSON blob when needed (e.g. when a new CVE is announced) to GCS (
  Google Cloud Bucket)
- Using Hugo and other tooling (such as Netlify), publish the list from this

>   JSON blob on official k8s website during `k/website` build

Author: Agree, this would be a great thing to explore before graduating from alpha to beta. I will take note of it so we don't miss it.

### Risks and Mitigations

#### JSON blob construction will fail

If the generation of the JSON blob listing known CVEs were to fail, downstream
jobs also fail. If blob construction fails, the failure will alert the owners of
this feature and we will take action as needed. If the failure can not be fixed
in a reasonable amount of time, the CVE feed will be stale until it is fixed. In
case of an urgent need from the community to update the vulnerabilities feed,
JSON blob will be manually updated via `gsutil` command.

#### Misuse of Auto-Refresh feature

Without proper filtering and control over who can label GitHub issues, the list
of CVEs can become a list with poor signal-to-noise ratio making the list
unusable.

For this purpose, the filtering is applied such that only issues that are marked
as `closed`

> will be part of the list. Also, additionally, the `official-cve-feed` label is a

Author: Makes sense. The feed could be modified to add such occurrences, manually assuming they are rare.

-->

[kubernetes.io]: https://kubernetes.io/

[kubernetes/enhancements]: https://git.k8s.io/enhancements

[kubernetes/kubernetes]: https://git.k8s.io/kubernetes

[kubernetes/website]: https://git.k8s.io/website

## Summary

Currently it is not possible to filter for issues or PRs that are related to
CVEs announced by kubernetes. This KEP addresses this concern by labelling this
issues or PRs with the new label **official-cve-feed** using the automation. The
in-scope issues are the closed issues for which there is a CVE ID and is
officially announced as a Kubernetes CVE by SRC in the past.

## Motivation

With the growing number of eyes on Kubernetes, the number of CVEs related to
Kubernetes have increased. Although most CVEs are regularly fixed that directly
or indirectly or transitively impact Kubernetes, there is no single place to
programmatically subscribe or pull the data of fixed CVEs, for the end users of
Kubernetes. Current options are either broken or incomplete.

An auto-refreshing CVE feed will allow end users to programmatically fetch the
list of CVEs and allow them to get the latest information from kubernetes
community.

### Goals

> Create a periodically auto-refreshing list of official Kubernetes CVEs

Reviewer: I'd make machine-readability an explicit goal. To some extent https://groups.google.com/g/kubernetes-security-announce already covers this in a non-machine readable form.

Author: IIUC, you mean the feed should be parseable in a well-recognized semi-structured format like JSON but not limited to JSON, correct?

Author: @tallclair okay to resolve the comment assuming my understanding above is correct? We are already publishing the blob in JSON

Author: Made this explicit in Goals section

for the KEP to move to `implementable` status and be included in the release.

In some cases, the questions below should also have answers in `kep.yaml`. This
is to enable automation to verify the presence of the review, and to reduce review
burden and latency.

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--
Why should this KEP _not_ be implemented?
-->

> ## Alternatives

Reviewer: Out of curiosity, did you consider scraping this from the official cvelist json? e.g. https://github.com/CVEProject/cvelist/blob/master/2020/8xxx/CVE-2020-8562.json

CVEs in the upstream repo should have the product name set to `Kubernetes` (https://github.com/CVEProject/cvelist/blob/master/2020/8xxx/CVE-2020-8562.json#L16)

It might be good to add this here.

Reviewer 2: That's a nice alternative, and I'm very pleased to learn about this official JSON list.

I notice that https://github.com/CVEProject/cvelist is described as “a pilot”.

I'm definitely onboard with us [re]using the CVE JSON format, now that I know there is one.

One thing that the whole set of proposed implementations as things stand give us is faster publication (minutes).
I'm also imagining that at some future point, the public entry in the Kubernetes vulnerability list could become the `refsource` for the CVE. Designing a way to publish that updated list / page ahead of the official CVE update avoids putting a cycle in what we could otherwise call the publication DAG.

Reviewer 2: I reckon we could provide at least a sketch of tool that an end-user could use to fetch updates to the official (pilot) feed, and filter for new updates that cover Kubernetes. If the CVE project's Git feed stops being a pilot, we could bless that tool with more recognition and maybe maintain it?

Author: I agree that exploring integrations with this tool / repo could be useful. I am leaning towards tackling this integration when we graduate KEP from alpha to beta.

Author: To resolve this thread, are you both okay if I add a note to consider the CVEProject in future scope but currently mark it as non-goal?

community.

### Goals

Create a periodically auto-refreshing machine-readable list of official
Kubernetes CVEs

### Non-Goals

- Triage and vulnerability disclosure: This will continue to be done by SRC
- Listing CVEs that are identified in build time dependencies and container
  images. Only official CVEs announced by the Kubernetes SRC will be published
  in the feed.
- Integration with [CVEProject](https://github.com/CVEProject/cvelist) may
  happen at a future stage but currently is not planned or scoped.

### User Stories (Optional)

#### Story 1

As a K8s end user, I want a list of CVEs with relevant information that I can
fetch programmatically, so I can understand when new CVEs are announced

#### Story 2

As a K8s End User, I want to use my browser to get a list of fixed CVEs, from
the official K8s website so that I can trust it as an authoritative source of
data through implicit trust offered by website certificate and domain name.

#### Story 3

As a K8s maintainer, I want to create a process that auto-updates CVE feed, when
SRC announces new CVEs such that I do not have to do extra work to maintain this
feed manually

> 

Reviewer 3: #### Story 4

As a K8s platform provider maintainer, I want to automatically to know if my kubernetes clusters are vulnerable to any of the CVE:s SRC have announced. I want to have a programmatically available API to parse this kind of data so I can easily provide it in to to my platform.

Author: Added

If the hash file matches, writing to bucket will be skipped, If hash file is
different writing to bucket, will be triggered.

## Design Details

The steps to implement this design will involve a prow job that:

1. Queries Github API for fixed official CVEs
1. Generates a JSON blob based on the query results
1. Writes the JSON blob to gcs-bucket if it is different than existing blob
1. Triggers the `k/website` build using netlify
   [build-hook](https://docs.netlify.com/configure-builds/build-hooks/). Secret
   token to trigger build is added as External Secret. See example for
   [snyk-token](https://github.com/kubernetes/k8s.io/pull/2222/files)
1. `k/website` build pulls the JSON blob from gcs bucket during website rebuild,
   pulling it from gcs-bucket into something like
   `https://kubernetes.io/security/official-cve-feed.json`
1. `k/website` renders the JSON blob as an HTML table for viewing the list of
   fixed CVEs from a browser at this location:
   `https://kubernetes.io/docs/reference/issues-security/official-cve-feed`
   and linked from this
   page: `https://kubernetes.io/docs/reference/issues-security/security/`

> 

Author: You will able to curl this like REST API: https://kubernetes.io/security/official-cve-feed.json to get the raw JSON

- [ ] (R) Graduation criteria is in place
  - [ ] (
    R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806)
    must be hit
    by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for
  publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to
  mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/

[kubernetes/enhancements]: https://git.k8s.io/enhancements

[kubernetes/kubernetes]: https://git.k8s.io/kubernetes

[kubernetes/website]: https://git.k8s.io/website

## Summary

Currently it is not possible to filter for issues or PRs that are related to

> CVEs announced by kubernetes. This KEP addresses this concern by labelling this

Reviewer 4: *Kubernetes

> CVEs announced by kubernetes. This KEP addresses this concern by labelling this

Reviewer 5: nit!
```suggestion
CVEs announced by kubernetes. This KEP addresses this concern by labelling these
```

issues or PRs with the new label **official-cve-feed** using the automation. The
in-scope issues are the closed issues for which there is a CVE ID and is
officially announced as a Kubernetes CVE by SRC in the past.

## Motivation

With the growing number of eyes on Kubernetes, the number of CVEs related to
Kubernetes have increased. Although most CVEs are regularly fixed that directly
or indirectly or transitively impact Kubernetes, there is no single place to
programmatically subscribe or pull the data of fixed CVEs, for the end users of

> Kubernetes. Current options are either broken or incomplete.

Author: Done

An auto-refreshing CVE feed will allow end users to programmatically fetch the

> list of CVEs and allow them to get the latest information from kubernetes

Reviewer 5: ```suggestion
list of CVEs and allow them to get the latest information from Kubernetes
```

community.

### Goals

Create a periodically auto-refreshing machine-readable list of official
Kubernetes CVEs

### Non-Goals

- Triage and vulnerability disclosure: This will continue to be done by SRC
- Listing CVEs that are identified in build time dependencies and container
  images. Only official CVEs announced by the Kubernetes SRC will be published
  in the feed.
- Integration with [CVEProject](https://github.com/CVEProject/cvelist) may
  happen at a future stage but currently is not planned or scoped.

### User Stories (Optional)

#### Story 1

As a K8s end user, I want a list of CVEs with relevant information that I can
fetch programmatically, so I can understand when new CVEs are announced

#### Story 2

As a K8s End User, I want to use my browser to get a list of fixed CVEs, from
the official K8s website so that I can trust it as an authoritative source of
data through implicit trust offered by website certificate and domain name.

#### Story 3

As a K8s maintainer, I want to create a process that auto-updates CVE feed, when
SRC announces new CVEs such that I do not have to do extra work to maintain this
feed manually

### Story 4

> As a K8s platform provider, I want to automatically to know if my kubernetes 

Reviewer 4: *Kubernetes

Reviewer 5: ```suggestion
As a K8s platform provider, I want to automatically know if my Kubernetes 
```

this feature and we will take action as needed. If the failure can not be fixed
in a reasonable amount of time, the CVE feed will be stale until it is fixed. In
case of an urgent need from the community to update the vulnerabilities feed,
JSON blob will be manually updated via `gsutil` command.

#### Misuse of Auto-Refresh feature

Without proper filtering and control over who can label GitHub issues, the list
of CVEs can become a list with poor signal-to-noise ratio making the list
unusable.

For this purpose, the filtering is applied such that only issues that are marked
as `closed`
will be part of the list. Also, additionally, the `official-cve-feed` label is a
[restricted](https://github.com/kubernetes/test-infra/blob/master/config/prow/plugins.yaml#L140-L150)
label that can only be applied by SRC and SIG Security Tooling Leads.

#### Large JSON blob could lead to slower read/write and resource consumption

Blobs will only be rewritten, if the generated blob is different from existing
blob. As hash file would be created and stored alongside generated blob. This
hash file will be check everytime before push to the hash of the generated file.

> If the hash file matches, writing to bucket will be skipped, If hash file is

Reviewer 4: second *if

## Design Details

The steps to implement this design will involve a prow job that:

1. Queries Github API for fixed official CVEs
1. Generates a JSON blob based on the query results
1. Writes the JSON blob to gcs-bucket if it is different than existing blob
1. Triggers the `k/website` build using netlify
   [build-hook](https://docs.netlify.com/configure-builds/build-hooks/). Secret
   token to trigger build is added as External Secret. See example for
   [snyk-token](https://github.com/kubernetes/k8s.io/pull/2222/files)
1. `k/website` build pulls the JSON blob from gcs bucket during website rebuild,
   pulling it from gcs-bucket into something like
   `https://kubernetes.io/security/official-cve-feed.json`
1. `k/website` renders the JSON blob as an HTML table for viewing the list of
   fixed CVEs from a browser at this location:
   `https://kubernetes.io/docs/reference/issues-security/official-cve-feed`
   and linked from this
   page: `https://kubernetes.io/docs/reference/issues-security/security/`

*Notes*:

- A GCS bucket needs to be created. Example PR for this looks
  like [this](https://github.com/kubernetes/k8s.io/pull/2570/files)

### Test Plan

> <!--
> **Note:** *Not required until targeted at a release.*
> 
> Consider the following in developing a test plan for this enhancement:
> - Will there be e2e and integration tests, in addition to unit tests?
> - How will it be tested in isolation vs with other components?
> 
> No need to outline all of the test cases, just the general strategy. Anything
> that would count as tricky in the implementation, and anything particularly
> challenging to test, should be called out.
> 
> All code is expected to have adequate tests (eventually with coverage
> expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
> when drafting this test plan.
> 
> [testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
> -->

Reviewer 6: Update to use the updated Test Plan section
```suggestion
<!--
<!--
**Note:** *Not required until targeted at a release.*
The goal is to ensure that we don't accept enhancements with inadequate testing.

All code is expected to have adequate tests (eventually with coverage
expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
when drafting this test plan.

[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[ ] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

<!--
Based on reviewers feedback describe what additional tests need to be added prior
implementing this enhancement to ensure the enhancements have also solid foundations.
-->

##### Unit tests

<!--
In principle every added code should have complete unit test coverage, so providing
the exact set of tests will not bring additional value.
However, if complete unit test coverage is not possible, explain the reason of it
together with explanation why this is acceptable.
-->

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
-->

- `<package>`: `<date>` - `<test coverage>`

##### Integration tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- <test>: <link to test coverage>

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

- <test>: <link to test coverage>
-->
```

An auto-refreshing CVE feed will allow end users to programmatically fetch the list of CVEs and allow them to get up to date information from kubernetes community.

### Goals {#goals}

Create a periodically auto-refreshing list of official Kubernetes CVEs

## Proposal {#proposal}

### Pre-requisites
- [x] https://github.com/kubernetes/test-infra/pull/23428
- [x] Search and Identify closed issues that have a CVE ID e.g. CVE-1001-12345 in the issue description or summary (This search [filter](https://github.com/kubernetes/kubernetes/issues?q=is%3Aissue+in%3Abody+%22CVSS%3A3.%22+label%3Acommittee%2Fsecurity-response+is%3Aclosed+) is giving the most accurate data so far)
- [x] Label those issues with `official-cve-feed` using https://docs.github.com/en/rest/reference/issues REST API
- [x] https://github.com/kubernetes/committee-security-response/pull/133

### Goals

- Generate a JSON document using the results from the filtered label on `k/k` repo.
- Create a Prow job to periodically generate this JSON document.
- Update the JSON doc when needed (e.g. when a new CVE is announced) in `k/website`
- Using Hugo, publish the list from this JSON document on official k8s website

> ### Non-Goals

Reviewer 7: Recommend moving this up to be under Goals in Motivation section.

- Triage and vulnerability disclosure: This will continue to be done by SRC
- CVEs that are identified in build time dependencies and container images. Only official CVEs announced by SRC will be published in the feed.

> ### User Stories (Optional)

Reviewer 7: ```suggestion
### User Stories
```

Reviewer 7: Recommend moving this to the top of the Proposal section, since it functions as a bridge between the very general Motivations / Goal and the more specific proposal.

## Summary {#summary}

Currently it is not possible to filter for issues or PRs that are related to CVEs announced by kubernetes. This KEP addresses this concern by labelling this issues or PRs with the new label **official-cve-feed** using the automation. The in-scope issues are the closed issues for which there is a CVE ID and is officially announced as a Kubernetes CVE by SRC in the past.

## Motivation {#motivation}

With the growing number of eyes on Kubernetes, the number of CVEs related to Kubernetes have increased. Although most CVEs are regularly fixed that directly or indirectly or transitively impact Kubernetes, there is no single place to programmatically subscribe or pull the data of fixed CVEs, for the end users of Kubernetes. Current options are either broken or incomplete.

An auto-refreshing CVE feed will allow end users to programmatically fetch the list of CVEs and allow them to get up to date information from kubernetes community.

### Goals {#goals}

Create a periodically auto-refreshing list of official Kubernetes CVEs

### Non-Goals

- Triage and vulnerability disclosure: This will continue to be done by SRC

> - CVEs that are identified in build time dependencies and container images. Only official CVEs announced by SRC will be published in the feed.

Reviewer 2: ```suggestion
- Listing CVEs that are identified in build time dependencies and container images. Only official CVEs announced by the Kubernetes SRC will be published in the feed.
```

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.
-->

#### Story 1

> As a K8s end user, I want a list of CVEs with relevant information that I can fetch programmatically, so I can understand when new CVEs are announced

Reviewer 2: Maybe also a story about acceding the HTML rendered list in a browser, and trusting that said page is current?

If that's not in scope, let's list it as a non-goal for _this_ KEP and explain that it's covered elsewhere. I hope it is in scope though.

Author: Added

#### Story 2

As a K8s maintainer, I want to create a process that auto-updates CVE feed, when SRC announces new CVEs such that I do not have to do extra work to maintain this feed manually

## Proposal {#proposal}

### Pre-requisites
- [x] https://github.com/kubernetes/test-infra/pull/23428
- [x] Search and Identify closed issues that have a CVE ID e.g. CVE-1001-12345 in the issue description or summary (This search [filter](https://github.com/kubernetes/kubernetes/issues?q=is%3Aissue+in%3Abody+%22CVSS%3A3.%22+label%3Acommittee%2Fsecurity-response+is%3Aclosed+) is giving the most accurate data so far)
- [x] Label those issues with `official-cve-feed` using https://docs.github.com/en/rest/reference/issues REST API
- [x] https://github.com/kubernetes/committee-security-response/pull/133

### Overview

- Generate a JSON document using the results from the filtered label on `k/k` repo.
- Create a Prow job to periodically generate this JSON document.
- Update the JSON doc when needed (e.g. when a new CVE is announced) in `k/website`

> - Using Hugo, publish the list from this JSON document on official k8s website

Reviewer 2: ```suggestion
- Using Hugo and other tooling (such as Netlify), publish the list from this JSON document on official k8s website
```

### Risks and Mitigations

#### JSON blob construction will fail

> If this happens, we expect the job too fail. If blob construction fails, the failure will alert the owners of this feature and we will take action as needed. If the failure can not be fixed in a reasonable amount of time, the CVE feed will be stale until it is fixed. In case of urgency from community to get the refreshed feed, JSON blob will be manually updated via usual PR review and approval process.

Reviewer 2: ```suggestion
If the generation of the JSON document listing known CVEs were to fail, downstream jobs also fail.
If blob construction fails, the failure will alert the owners of this feature and we will take action as needed. If the failure can not be fixed in a reasonable amount of time, the CVE feed will be stale until it is fixed. In case of an urgent need from the community to update the vulnerabilities feed, JSON blob will be manually updated via usual PR review and approval process.
```

#### Urgent CVE feed refresh

In some extenuating circumstances, we may need to update the CVE feed within minutes of the official CVE announcement, instead of waiting for the merge based or periodical website rebuild. In those situations, manual updates to JSON blob using usual PR reviews and approval process can be implemented.

#### Misuse of Auto-Refresh feature

Without proper filtering and control over who can label GitHub issues, the list of CVEs can become a list with poor signal to noise ratio making the list unusable. 

For this purpose, the filtering is applied such that only issues that are marked as `closed`
will be part of the list. Also, additionally, the `official-cve-feed` label is a 
[restricted](https://github.com/kubernetes/test-infra/blob/master/config/prow/plugins.yaml#L140-L150)

> label that can only be applied by SRC and SIG Security Tooling Leads.

Reviewer 2: Maybe add some text about why closed issues are the right ones to list.
(I'd expect to mention all labelled issues, closed or not, provided that they are publicly visible).

Author: My assumption is that open SRC created GitHub issues with a CVE ID that are not under embargo, are unlikely. Maybe worth adding that as an assumption as a clarification?

#### Storage of CVE feed blob

There are two options to store the CVE feed JSON blob:
* __Google Cloud Bucket__: A new google cloud bucket can be created where the CVE feed is written and read using `gsutil` tool.
* __Git repository__: Store it as a version controlled artifact in one of the kubernetes org websites.

Google cloud option has an advantage of transparent updates to JSON blob where the prow job run will be identical everytime. The disadvantage with Google Cloud bucket is that, it will have an unofficial looking URL which would be hard for an end user to deciper it's authenticity and provenance.

> Advantage with git repository especially `k/website` hosts the JSON blob the domain name in the URL would be `k8s.io/static/security/official-cve-feed.json` which is much more recognizable, intuitive in terms of trust, TLS enabled and unlikely to be spoofed. The disadvantage though is that this might get delayed by PR review and approval process. However, this can be prevented through use of `skip-review` label.

Reviewer 2: I'd expect to see [the choice we didn't select] mentioned under the heading “Alternatives”, later in the KEP.

Author: Refactored under alternatives

data through implicit trust offered by website certificate and domain name.

#### Story 3

As a K8s maintainer, I want to create a process that auto-updates CVE feed, when
SRC announces new CVEs such that I do not have to do extra work to maintain this
feed manually

## Proposal {#proposal}

### Pre-requisites

- [x] Add official-cve-label https://github.com/kubernetes/test-infra/pull/23428
- [x] Search and Identify closed issues that have a CVE ID e.g. CVE-1001-12345
  in the issue description or summary (This
  search [filter](https://github.com/kubernetes/kubernetes/issues?q=is%3Aissue+in%3Abody+%22CVSS%3A3.%22+label%3Acommittee%2Fsecurity-response+is%3Aclosed+)
  is giving the most accurate data so far)
- [x] Label those issues with `official-cve-feed`
  using https://docs.github.com/en/rest/reference/issues REST API
- [x] Add `official-cve-feed` label as part of SRC
  playbook: https://github.com/kubernetes/committee-security-response/pull/133

### Overview

- Generate a JSON document using the results from the filtered label on `k/k`
  repo.
- Create a Prow job to periodically generate this JSON document.

> - Update the JSON doc when needed (e.g. when a new CVE is announced)

Reviewer 8: does the file need to be in k/website specifically?
if we had a repo scoped to this task, a robot account could probably be allowed to push directly to the repo, eliminating the need for an out-of-org fork.

Author: Yes it needs to be in `k/website`, because as is described in the rest of the KEP, it allows us to host this data under `https://k8s.io/static/official-cve-list.json` which grants a higher level of (explicit) trust compared to say something like `https://raw.githubusercontent.com/kubernetes/sig-security/main/official-cve-list.json`

Reviewer 9: I'm fairly certain the netlify/hugo config can aggregate content from other sources into the website. Have you talked with sig-docs? I know we do some of this for the kubernetes/contributor-site repo to pull in stuff that's in sig-release and community repos.

Author: @cblecker If that's possible happy to take a look. @sftim who has context about this KEP from SIG Docs to see if that is an option and to share any specific examples that do this.

Reviewer 2: > https://k8s.io/static/official-cve-list.json

It would be https://k8s.io/SOMETHING_ELSE/official-cve-list.json I think

Reviewer 2: The other benefit, I just remembered, is that if you push to the _main_ branch of k/website, that triggers a build. If you push anywhere else, you still need a way to trigger a build for k/website.

Author: Yes, @sftim is spot on. Unless we find a way to trigger a build of `k/website` without a PR to `k/website` I will add a new section in alternatives considered to add in the KEP :)

Reviewer 9: Considering the production site is rebuilt multiple times a day every day, is instant availability of updates to this feed an explicit requirement (and worth the overhead of running another bot that would need to be maintained)?

Reviewer 2: I've assumed that it's useful to have the list of CVEs current when we announce risks (and I do think it looks bad to announce a vulnerability when that _isn't_ then shown on the website for another 12 hours - which is the worst case).

Author: Thanks for all the feedback. KEP PR is now updated, with the new flow: push to k/sig-security and pull the feed to k/website triggered via build hook. The reasons for not pushing to k/website directly are also mentioned under alternatives considered. Specific commit that addresses these changes is here: https://github.com/kubernetes/enhancements/pull/3204/commits/ebf89c6b6d98fb403b35b02c9925d02fb3d082c4

Relevant slack thread where we are discussing this: https://kubernetes.slack.com/archives/C09QZ4DQB/p1645563124641899?thread_ts=1645129435.563709&cid=C09QZ4DQB
